File: day-1/php/main.php
<?php

function getTotal($array) {
	$sum = 0;
	ksort($array);

	$firstKey = array_key_first($array);
	$lastKey = array_key_last($array);

	if (isset($array[$firstKey]) || isset($array[$lastKey])) {
		$sum = $array[$firstKey] . $array[$lastKey];
	}

	return $sum;
}

$input = fopen("../input.txt", "r") or die("Unable to open file!");
$search = [
	1 => ["1", "one"],
	2 => ["2", "two"],
	3 => ["3", "three"],
	4 => ["4", "four"],
	5 => ["5", "five"],
	6 => ["6", "six"],
	7 => ["7", "seven"],
	8 => ["8", "eight"],
	9 => ["9", "nine"],
];
$one_star_result = 0;
$two_star_result = 0;

while(!feof($input)) {
	$line = fgets($input);
	$one_star_array = [];
	$two_star_array = [];

	foreach ($search as $key => $value) {
		foreach ($value as $search_key => $search_value) {
			$first = strpos($line, $search_value);
			$last = strrpos($line, $search_value);
			if ($first !== false) {
				if ($search_key == 0) {
					$one_star_array[$first] = $key;
				}
				$two_star_array[$first] = $key;
			}
			if ($last !== false) {
				if ($search_key == 0) {
					$one_star_array[$last] = $key;
				}
				$two_star_array[$last] = $key;
			}
		}
	}

	$one_star_total = getTotal($one_star_array);
	$one_star_result += $one_star_total;

	$two_star_total = getTotal($two_star_array);
	$two_star_result += $two_star_total;
}

fclose($input);

echo 'One star result: ' . $one_star_result;
echo "\n";
echo 'Two star result: ' . $two_star_result;

?>

File: day-1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	search := map[int][]string{
		1: {"1", "one"},
		2: {"2", "two"},
		3: {"3", "three"},
		4: {"4", "four"},
		5: {"5", "five"},
		6: {"6", "six"},
		7: {"7", "seven"},
		8: {"8", "eight"},
		9: {"9", "nine"},
	}

	oneStarResult := 0
	twoStarResult := 0

	file, err := os.Open("../input.txt")
	if err != nil {
		log.Fatal(err)
	}

	reader := bufio.NewReader(file)

	for {
		line, _, err := reader.ReadLine()
		oneStarMap := make(map[int]int)
		twoStarMap := make(map[int]int)

		if len(line) > 0 {
			for key, value := range search {
				for searchKey, searchValue := range value {

					first := strings.Index(string(line), searchValue)
					last := strings.LastIndex(string(line), searchValue)

					if first != -1 {
						if searchKey == 0 {
							oneStarMap[first] = key
						}
						twoStarMap[first] = key
					}

					if last != -1 {
						if searchKey == 0 {
							oneStarMap[last] = key
						}
						twoStarMap[last] = key
					}
				}
			}

		}
		if err != nil {
			break
		}

		oneStarTotal := getTotal(oneStarMap)
		oneStarResult += oneStarTotal

		twoStarTotal := getTotal(twoStarMap)
		twoStarResult += twoStarTotal
	}

	fmt.Println("1 Star Result:", oneStarResult)
	fmt.Println("2 Star Result:", twoStarResult)
}

func getTotal(starMap map[int]int) int {
	sum := 0

	if len(starMap) > 0 {
		keys := []int{}

		for key := range starMap {
			keys = append(keys, key)
		}

		sort.Ints(keys)

		firstStr := strconv.Itoa(starMap[keys[0]])
		lastStr := strconv.Itoa(starMap[keys[len(keys)-1]])
		strConcat := firstStr + lastStr

		newSum, err := strconv.Atoi(strConcat)
		if err != nil {
			panic(err)
		}

		sum = newSum
	}

	return sum
}
